### ISSUE TITLE
Can't sync users with Active Directory

### ISSUE THREAD

**Issue by beatcracker** [dnaeon/zabbix-ldap-sync#16]
I'm trying to sync AD group with Zabbix, but I'm running into some issues. Here is my group/user(s):

###### Group:
```
PS > dsquery group -name "ABC Development"
"CN=ABC Development,CN=Users,DC=XYZ,DC=mycompany,DC=com"
```

###### One of the users in the group:
```
PS > dsquery user -samid "johndoe"
"CN=John Doe,OU=Software Development,DC=XYZ,DC=mycompany,DC=com"
```

If I use DN: `DC=XYZ,DC=mycompany,DC=com`, script creates group in Zabbix, but doesn't add any users to it. If I use this DN: `CN=Users,DC=XYZ,DC=mycompany,DC=com`, script fails with error:
```
    dn, users = uid.pop()
IndexError: pop from empty list
```

I've modified script to dump some debug info:

```python
        result = self.conn.search_s(base=self.base,
                                    scope=ldap.SCOPE_SUBTREE,
                                    filterstr=filter,
                                    attrlist=attrlist)

        print 'get_group_memebers result: %s' % result

        if not result:
            print '>>> Unable to find group %s, skipping group' % group
            return None

        # if this is activedirectory, we have to replace each CN with the actual username
        if active_directory:

            final_listing = []

            for members in result:
                result_dn = members[0]
                result_attrs = members[1]

            if "member" in result_attrs:
                for member in result_attrs["member"]:

                    attrlist = ['sAMAccountName']
                    filter   = '(&(objectClass=person)(distinguishedName=%s))' % ldap.filter.escape_filter_chars(member)

                    # get the actual LDAP object for each group member

                    print 'base: %s' % self.base
                    print 'ldap.SCOPE_SUBTREE: %s' % ldap.SCOPE_SUBTREE
                    print 'filter: %s' % filter
                    print 'attrlist: %s' % attrlist

                    uid = self.conn.search_s(base=self.base,
                                    scope=ldap.SCOPE_SUBTREE,
                                    filterstr=filter,
                                    attrlist=attrlist)

```

Here is debug ouput for both  `DN`'s:
<hr/>
###### Config file (DC=XYZ,DC=mycompany,DC=com)
```
[ldap]
type = activedirectory
uri = ldap://10.22.33.44:389/
base = DC=XYZ,DC=mycompany,DC=com
binduser = XYZ\winbind
bindpass = asdfgh
groups = ABC Development

[zabbix]
server = http://localhost/zabbix/
username = ldapsync
password = qwerty
```
###### Debug output
```
[root@ZABBIX zabbix-ldap-sync]# ./zabbix-ldap-sync -f ./zabbix-ldap.conf

>>> Creating Zabbix group ABC Development
>>> Group ABC Development created with groupid 24

get_group_memebers result: [('CN=ABC Development,CN=Users,DC=XYZ,DC=mycompany,DC=com', {'member': ['CN=John Doe,OU=Software Development,DC=XYZ,DC=mycompany,DC=com', 'CN=Jane Doe,OU=Software Development,DC=XYZ,DC=mycompany,DC=com']}), (None, ['ldap://ForestDnsZones.XYZ.mycompany.com/DC=ForestDnsZones,DC=XYZ,DC=mycompany,DC=com']), (None, ['ldap://DomainDnsZones.XYZ.mycompany.com/DC=DomainDnsZones,DC=XYZ,DC=mycompany,DC=com']), (None, ['ldap://XYZ.mycompany.com/CN=Configuration,DC=XYZ,DC=mycompany,DC=com'])]
```
<hr/>
###### Config file (CN=Users,DC=XYZ,DC=mycompany,DC=com)
```
[ldap]
type = activedirectory
uri = ldap://10.22.33.44:389/
base = CN=Users,DC=XYZ,DC=mycompany,DC=com
binduser = XYZ\winbind
bindpass = asdfgh
groups = ABC Development

[zabbix]
server = http://localhost/zabbix/
username = ldapsync
password = qwerty
```

###### Debug output
```
[root@ZABBIX zabbix-ldap-sync]# ./zabbix-ldap-sync -f ./zabbix-ldap.conf

>>> Creating Zabbix group ABC Development
>>> Group ABC Development created with groupid 22

get_group_memebers result: [('CN=ABC Development,CN=Users,DC=XYZ,DC=mycompany,DC=com', {'member': ['CN=John Doe,OU=Software Development,DC=XYZ,DC=mycompany,DC=com', 'CN=Jane Doe,OU=Software Development,DC=XYZ,DC=mycompany,DC=com']})]

base: CN=Users,DC=XYZ,DC=mycompany,DC=com

ldap.SCOPE_SUBTREE: 2

filter: (&(objectClass=person)(distinguishedName=CN=John Doe,OU=Software Development,DC=XYZ,DC=mycompany,DC=com))

attrlist: ['sAMAccountName']

Traceback (most recent call last):
  File "./zabbix-ldap-sync", line 543, in <module>
    main()
  File "./zabbix-ldap-sync", line 540, in main
    zabbix_conn.sync_users(config.ldap_uri, config.ldap_base, config.ldap_groups, config.ldap_user, config.ldap_pass, config.ldap_mail)
  File "./zabbix-ldap-sync", line 418, in sync_users
    ldap_users = ldap_conn.get_group_members(eachGroup)
  File "./zabbix-ldap-sync", line 128, in get_group_members
    dn, users = uid.pop()
IndexError: pop from empty list
```
<hr/>

I'm running this in `CentOS release 6.7 (Final)` with [Python 2.7 via SCL](https://www.softwarecollections.org/en/scls/rhscl/python27/)

What can I do to sync groups succesfully?

**Comment by ygotame** [comment]
Hello @beatcracker  did you manage to get up update that issue ? 
@dnaeon  Do you have any idea ?

When I tried with port 389, I do not get nothing output. Via Web, I did configure LDAP with AD. 

Therefore, I have having same issue too. @beatcracker  did you also tried with port 3268

`raceback (most recent call last):
  File "./zabbix-ldap-sync", line 536, in <module>
    main()
  File "./zabbix-ldap-sync", line 533, in main
    zabbix_conn.sync_users(config.ldap_uri, config.ldap_base, config.ldap_groups, config.ldap_user, config.ldap_pass, config.ldap_mail)
  File "./zabbix-ldap-sync", line 411, in sync_users
    ldap_users = ldap_conn.get_group_members(eachGroup)
  File "./zabbix-ldap-sync", line 121, in get_group_members
    dn, users = uid.pop()
IndexError: pop from empty list
`

`[ldap]
type = activedirectory
uri = ldap://1.1.1.1:3268
base = DC=SERVER-LOCAL,DC==COM
binduser = SERVER-LOCAL\zabbix
bindpass = Pass*******
groups = Zabbix-Users,Zabbix-Admin

[zabbix]
server = http://1.1.1.2/zabbix/
username = adminzab
password = Pass*******
`

**Comment by beatcracker** [comment]
@orymate This is probably related to user being in specific `OU`.

 I've had limited success by replacing in `get_group_members`

                    uid = self.conn.search_s(base=self.base,
                                    scope=ldap.SCOPE_SUBTREE,
                                    filterstr=filter,
                                    attrlist=attrlist)
with

    uid = self.conn.search_s(base=member,
                    scope=ldap.SCOPE_SUBTREE,
                    attrlist=attrlist)
.

But `get_user_email` fails to get emails, because resulting filter `'(&(objectClass=person)(sAMAccountName=johndoe))'` wouldn't match anything.

**Comment by ygotame** [comment]
@beatcracker  thank's that's working for me too. Still doing some tweak (like assign user automatically to super-admin)
What is good is that I use port 3268 (which look over whole domain), therefore I am getting all the info of the user as well as email too. It is automatically added to media type. But when I use, 389 it is not working well neither the login for some users because they are listed in nest group  which is not inclued in Zabbix OU.

For info about the port : 

Port 3268. This port is used for queries specifically targeted for the global catalog. LDAP requests sent to port 3268 can be used to search for objects in the entire forest. However, only the attributes marked for replication to the global catalog can be returned. For example, a user’s department could not be returned using port 3268 since this attribute is not replicated to the global catalog.

Port 389. This port is used for requesting information from the local domain controller. LDAP requests sent to port 389 can be used to search for objects only within the global catalog’s home domain. However, the requesting application can obtain all of the attributes for those objects. For example, a request to port 389 could be used to obtain a user’s department.

**Comment by beatcracker** [comment]
I'm kinda fixed this: now full `distinguished name` is used everywhere for lookups.

Works for me, but it breaks nix LDAP compatibility, because I have no idea what those LDAP queries would return. Should be easy to fix for someone with access to nix LDAP implementation.

If somone is willing to try it, grab my version here: https://github.com/beatcracker/zabbix-ldap-sync/tree/ad_fix

**Comment by beatcracker** [comment]
@ygotame 

> Still doing some tweak (like assign user automatically to super-admin)

I've just added abilty to set [custom properties](https://www.zabbix.com/documentation/3.2/manual/api/reference/user/object) for users in group via additional `[user]` section in config file. In preliminary tests it works fine for `type`, `url` and `atologin`.

**Comment by dnaeon** [comment]
@beatcracker, @ygotame

Sorry, for not following this issue closely, but I lost my dev Zabbix environment, so I cannot really check this.

Feel free to send patches in case you get this one fixed and I'll make sure to merge them.

Thanks,
Marin

**Comment by ygotame** [comment]
@beatcracker @dnaeon 
cool guys. I will test on my dev Zabbix and let you know the update .After we can close the issue and create a new one for other purpose.

**Comment by beatcracker** [comment]
@dnaeon Thanks, can you recommend a Linux LDAP VM appliance that I can spin up quickly to fix nix LDAP codepath? Is this one ok: https://www.turnkeylinux.org/openldap ?

**Comment by dnaeon** [comment]
@beatcracker, I haven't really used an LDAP appliance myself, so cannot recommend one. I usually just go with the LDAP packages.

The link you've provided seems would do the job the job as well, so I guess you would be okay to go with it.

Thanks,
Marin

**Comment by ygotame** [comment]
@beatcracker  Before, I was using that application too but have you consider Fusion Directory https://www.fusiondirectory.org for web interface.

About running fast VM, you can use Docker container which is easy to install and run right a way for test and dev :https://github.com/osixia/docker-openldap

**Comment by dnaeon** [comment]
Fixed in #21
